### ISSUE TITLE
결승전 발표 후 질문 - sherlock

### ISSUE THREAD

**Issue by Kang-heesuk** [D2CAMPUS-FEST/2016#21]
* repo : https://github.com/mjkim610/sherlock

**Comment by doortts** [comment]
셜록 - 핑거프린트
---
- 등록된 이슈가 거의 없던데 기능 개발은 어떤 걸 기준으로 하나요? 
- 성장과 확산에 가장 큰 걸림돌은 무엇이라고 생각하고 있으신가요?
- 점수의 기준에 대한 의미? 이를테면 70과 80의 차이가 설정하는 사람 입장에서는 유의미하다고 느끼기 어려울 것 같은데요.
- 공용 PC의 경우는 사용성이 어떻게 되나요?


- Contribution 하고 싶은 사람을 위한 내용이 없어서 아쉽네요.
- 유사 시스템에 대한 대체 가능 예제가 있으면 좋겠어요

**Comment by hahnlee** [comment]
로그인 시도 리미트가 있나요?

**Comment by mjkim610** [comment]
@doortts 
- 등록된 이슈가 거의 없던데 기능 개발은 어떤 걸 기준으로 하나요?
    - 저희는 대부분의 프로젝트 진행 방향을 오프라인 모임을 통해 사전 협의를 했기 때문에 깃헙에 정리를 하지 못 했습니다. 오픈소스 프로젝트인 만큼 앞으로 꾸준히 기록을 하도록 하겠습니다. 좋은 지적 감사합니다!

- 성장과 확산에 가장 큰 걸림돌은 무엇이라고 생각하고 있으신가요?
    - 아무래도 인증 시스템이다 보니 보안에 대한 100%의 신뢰도를 확보하는 것이 중요하다고 생각합니다. 여기서 신뢰도는 기술적 신뢰도 및 셜록에 대한 고객(?)들의 신뢰도 모두 생각해야 한다고 생각합니다.

- 점수의 기준에 대한 의미? 이를테면 70과 80의 차이가 설정하는 사람 입장에서는 유의미하다고 느끼기 어려울 것 같은데요.
   - 수집한 로그를 분석해 browser fingerprint의 feature들을 백점만점의 점수형태로 환산합니다. 서비스 제공자는 이런 지표를 참고할 수 있지만, 셜록은 threshold 기준을 정하는 방법을 제공하는 것이고 실제로 기준을 정하는 것은 서비스 제공자가 본인의 서비스에 대한 판단을 통해 하는 것입니다. 향후 데이터가 더 수집된다면 보안 수준 상/중/하 에 대한 템플릿을 제공할 수 있을 것 같습니다.
![image](https://cloud.githubusercontent.com/assets/17107206/23053774/0a8b1ba2-f51f-11e6-99f6-c64aa5f64286.png)


- 공용 PC의 경우는 사용성이 어떻게 되나요?
    - 공용 PC에서는 셜록의 사용을 권하지 않습니다. 공용 PC는 사용자가 자주 사용하지 않을 뿐만 아니라 여러 사람이 사용할 수 있기 때문에 browser fingerprint를 기록하는 것이 부적절하다고 판단합니다. 마치 비밀번호를 12345로 설정하는 것 과 다름 없다고 생각합니다. 

- Contribution 하고 싶은 사람을 위한 내용이 없어서 아쉽네요.
    - Pull request는 언제나 환영입니다! 조만간 README.md 에 contribution 안내 항목을 추가하겠습니다.

- 유사 시스템에 대한 대체 가능 예제가 있으면 좋겠어요
    - 유사 시스템이라함은 슬랙의 Magic Link, 기타 로그인 시스템의 Fraud Detection System을 말씀하시는 건가요? 저희도 다른 인증 시스템과의 속도 비교는 간단하게 해봤습니다. 앞으로 더 심도 있는 연구를 하겠습니다!
![image](https://cloud.githubusercontent.com/assets/17107206/23053786/1978f0f8-f51f-11e6-8536-54c3f93ddc1e.png)


좋은 의견 감사합니다!

**Comment by kim-sardine** [comment]
@sn0wle0pard 횟수의 리미트 말씀이신가요? 현재는 존재하지 않습니다. 그래도 현재 모든 로그인시도를 로그로 남기고있기 때문에 단시간내 다수의 로그인 시도가 발생할때 블럭하는 방법을 고민중입니다.

**Comment by hahnlee** [comment]
@restforest 그렇군요 페이스북도 개발자 페이지에 비밀번호 찾기 시도 제한이 없던 보안 이슈가 있어서 문제가 되었던 적이 있는데 블럭 기능이 있으면 보안에 도움 될 것 같습니다

**Comment by UnoRich** [comment]
개인피씨를 다른사라이 사용하여 메일을 아는경우에는 위험 할 수 있을거같은데요. 접속시에 접속흔적을 노티피케이션해주면 좋을것 같습니다.

**Comment by Luavis** [comment]
프로젝트 특성상 개인 PC에 포커스를 두어 개발된거 같은데 일반적인 웹 사이트에서 지원하는 자동로그인이나 Auto fill up과 차별점이 있을까요? 오히려 로그인 하지 않아도 이미 로그인 되어있는 환경의 경우가 더 편하지 않나요?

**Comment by raon0211** [comment]
Browser Fingerprint는 어떤 사이트나 쉽게 수집할 수 있는데요, 오픈 소스라는 환경 특성 상 보안에 사용되는 Fingerprint의 종류는 공개되어 있는 현실이에요... 그렇다면 악의적인 사이트가 사용자의 Browser Fingerprint를 수집하여 해킹 공격에 사용할 수 있지 않을까요? 이러한 부분을 어떻게 보완할 수 있을까요?

**Comment by mjkim610** [comment]
@Luavis 
- 프로젝트 특성상 개인 PC에 포커스를 두어 개발된거 같은데 일반적인 웹 사이트에서 지원하는 자동로그인이나 Auto fill up과 차별점이 있을까요? 오히려 로그인 하지 않아도 이미 로그인 되어있는 환경의 경우가 더 편하지 않나요?
    - 이미 로그인 되어있는 경우는 autofill이 아닌 세션쿠키를 활용한 것입니다. 웹서비스가 https로 잘 구현되어 있다면 MITM 공격은 막을 수 있지만 최근에 [PoisonTap](https://github.com/samyk/poisontap)과 같은 공격의 경우 세션 쿠키를 탈취해서 victim의 계정으로 로그인을 할 수 있는 공격이 있어서 셜록을 고안해낸 이유도 있습니다. 하지만 지적해주신대로 셜록이나 autofill이나 사용자가 체감하는 차이점은 없을 것입니다! 다만 셜록의 경우 browser fingerprint의 미묘한 차이에 대한 학습을 통해 보안성은 유지하되 사용성을 확장시키는 방법에 대해서 꾸준히 고민하겠습니다! (예를 들어, 서로 다른 기기이지만 같은 게이트웨이를 사용하기 때문에 같은 IP주소를 가지고 있을 경우, 동일 인물이 로그인을 시도하는 것으로 판단할 수 있을지에 대한 연구가 필요합니다.)

**Comment by mjkim610** [comment]
@raon0211 

- Browser Fingerprint는 어떤 사이트나 쉽게 수집할 수 있는데요, 오픈 소스라는 환경 특성 상 보안에 사용되는 Fingerprint의 종류는 공개되어 있는 현실이에요... 그렇다면 악의적인 사이트가 사용자의 Browser Fingerprint를 수집하여 해킹 공격에 사용할 수 있지 않을까요? 이러한 부분을 어떻게 보완할 수 있을까요?
    - 좋은 지적이라고 생각합니다. Fingerprint의 각 feature를 숨기는 것은 security by obscurity인 옳지 않은 보안 방식이기 때문에 차라리 강력한 서버사이드 암호화 프로토콜을 사용하고, 그 프로토콜과 salt값을 숨기는게 맞다고 생각합니다.

**Comment by kim-sardine** [comment]
@uenieng 네 그 부분은 자동로그인이나 자동완성에서부터 계속 문제가 되어오고있는것 같습니다.. . 더 현명하게 해결할수 있는 방안을 고민해보겠습니다. 의견 감사합니다!
